src/main/java/org/extism/chicory/sdk/Linker.java

```diff
@@ -0,0 +1,141 @@
+package org.extism.chicory.sdk;
+
+import com.dylibso.chicory.log.Logger;
+import com.dylibso.chicory.runtime.HostFunction;
+import com.dylibso.chicory.runtime.HostImports;
+import com.dylibso.chicory.runtime.Instance;
+import com.dylibso.chicory.runtime.Module;
+import com.dylibso.chicory.wasi.WasiOptions;
+import com.dylibso.chicory.wasi.WasiPreview1;
+
+/**
+ * Links together the modules in the given manifest with the given host functions
+ * and predefined support modules (e.g. the {@link Kernel}.
+ * <p>
+ * Returns a {@link LinkedModules} instance, which in turn can be converted
+ * into a {@link Plugin}.
+ */
+class Linker {
+    public static final String EXTISM_NS = "extism:host/env";
+    private final Manifest manifest;
+    private final HostFunction[] hostFunctions;
+    private final Logger logger;
+
+    Linker(Manifest manifest, HostFunction[] hostFunctions, Logger logger) {
+        this.manifest = manifest;
+        this.hostFunctions = hostFunctions;
+        this.logger = logger;
+    }
+
+    /*
+     * Try to find the main module:
+     *  - There is always one main module
+     *  - If a Wasm value has the Name field set to "main" then use that module
+     *  - If there is only one module in the manifest then that is the main module by default
+     *  - Otherwise the last module listed is the main module
+     *
+     */
+    public LinkedModules link() {
```

Comment:
I'm not how applicable this will be to the chicory-sdk, but! In the js-sdk, we split up the [manifest resolution](https://github.com/extism/js-sdk/blob/main/src/manifest.ts#L102) and [linking steps](https://github.com/extism/js-sdk/blob/main/src/foreground-plugin.ts#L155). Calling out some points of interest:

- The linker is recursive (so we visit the imports the main module, then the imports of those imports, etc etc.)
- WASIp1 command modules (and their subtree) are instantiated at _every_ point where they're imported.
- In the js-sdk, we allow circular imports of functions, but not of the other export types – JS doesn't give us a straightforward way to generate and inject table/memory/global resources, unfortunately.

Reply:
I think the recursive import resolution is a nice way to handle this - that also allows you to skip instantiating modules that aren't actually needed.

Reply:
interesting, so Manifest handling is already [handled](https://github.com/extism/chicory-sdk/blob/main/src/main/java/org/extism/chicory/sdk/ManifestWasm.java) [separately](https://github.com/extism/chicory-sdk/blob/main/src/main/java/org/extism/chicory/sdk/Manifest.java#L28); 

There are also two separate phases in the linker; i.e. the Linker here only resolves the modules (sure, not recursively) but won't instantiate. Instead, it returns a `LinkedModule` and then `LinkedModule#toPlugin` instantiates all `Module`(`.Builder`s) at once. 

Indeed, I am not cross-linking imports though 😬  I am quite sure that at least part of this logic belongs to Chicory itself 🤔

Reply:
followed up to this original work rebasing the work on the Store that has been introduced with https://github.com/dylibso/chicory/pull/495

this is still unfinished, at the bare minimum it lacks proper testing; however we can already see that the Kernel is now instantiated as a consequence of the requirement in the `main` module, instead of being provided regardless (not necessarily the desired behavior, but at least it shows that it works for this simple example :D)